### PR TITLE
[AppBar] Get rid of extra space for status bar when app bar is in popover

### LINKED ISSUE #4104: [AppBar] Header height is accounting for the status bar in modal sheets
From the bug filer:
> MDCAppBar can be either used as the screen's top bar or as a header in a form sheet. While for the first case the app bar extends its height to include the status bar, the latter one results in extra top padding that might look weird and affects the vertical alignment of the text relative to the header's center. Is the MDCAppBar view object intended to be of the same height for both cases? Does the text vertical alignment in the form sheet header conform to the guidelines?
>
> See example screenshot from the MDC Catalog app on the iPad: App Bar -> Modal Presentation -> Detail

simulator screen shot sep 6 2017 3 29 22 pm

---

This is an internal issue. If you are a Googler, please visit [b/65403467](http://b/65403467) for more details.

### PR DESCRIPTION
**Pull request by andrewoverton**
This PR closes #4104.

The first approach I took can be seen in 8715c0a. I think I like this one better though, because it doesn't add to `MDCLayoutMetrics`.

### CODE REVIEW

**Comment by ianegordon** [comment]
Can we see a before and after screenshot?
Please verify behavior on iPhone X and non-iPhone X to ensure safe areas are working as expected in both cases.
Please verify call in progress (double height status bar) works as expected.

File: components/FlexibleHeader/src/MDCFlexibleHeaderView.m [hunk 1]
```diff
@@ -1086,6 +1087,17 @@ - (void)fhv_contentOffsetDidChange {
   [self fhv_updateLayout];
 }
 
+- (CGFloat)additionalSpaceNeededForDeviceTopSafeAreaInset {
+  return [self isWithinDeviceTopSafeAreaInsets] ? 0 : MDCDeviceTopSafeAreaInset();
+}
+
+- (BOOL)isWithinDeviceTopSafeAreaInsets {
+  UIWindow *keyWindow = [UIApplication mdc_safeSharedApplication].keyWindow;
```
Comment (AlimoRabbani) [inline components/FlexibleHeader/src/MDCFlexibleHeaderView.m]: Is it possible to have an approach that also works in extensions? I believe sharedApplication is unavailable in extensions.

**Comment by andrewoverton** [comment]
I need to take another look at this... the "call in progress" green bar messes everything up :(

**Comment by andrewoverton** [comment]
Here is a before:
simulator screen shot - ipad air 2 - 2018-06-28 at 14 46 39
And an after:
simulator screen shot - ipad air 2 - 2018-06-28 at 14 46 04

@ianegordon, the in-call status bar was causing some issues, but they're resolved.
@AlimoRabbani, with the changes in 7b5c019 extension behavior will not be affected.

**Comment by andrewoverton** [comment]
Tested on various combinations of iOS 8 and 11 with iPhone 6/7/8/X and iPad 2.

**Comment by AlimoRabbani** [comment]
Right! This PR shouldn't affect extension behavior, but extension behavior is also currently broken because of a very similar reason to why app bar popover heights are broken (b/65403467 for the internal discussion). I was wondering if it's possible to have a fix that works in both cases.

If not, do we know how we can extend this PR in the future to tackle the extension issue?

**Comment by andrewoverton** [comment]
Ohhhh my bad, I didn't realize extension behavior was broken. Getting rid of that "if !extension" clause that I added may actually address the bad extension behavior. I'm not sure how I would test this though. Do you have any ideas @AlimoRabbani ? You don't happen to have a watch project that would be a good candidate for this do you?

**Comment by AlimoRabbani** [comment]
I can try patching it in and test it in my Share Extension, but that'll have to wait until tomorrow I think. Is that ok?

**Comment by andrewoverton** [comment]
Sure that’s alright @AlimoRabbani . Sorry to have you go through the trouble!

My gut is telling it me that the fix as it stands right now won’t address extension behavior. I think changing [this line](https://github.com/material-components/material-components-ios/blob/develop/components/private/UIMetrics/src/MDCLayoutMetrics.m#L46) to `CGFloat topInset = [[UIApplication mdc_safeSharedApplication] mdc_isAppExtension] ? 0 : kFixedStatusBarHeightOnPreiPhoneXDevices;` _might_ make a difference though?

**Comment by andrewoverton** [comment]
These changes fix the issue -- but not for extensions. They just feel brittle. So I'm going to clooooose the PR and have Jeff take a look at the changes when he gets back. He knows things about this suite of classes that may come in handy. Maybe he'll be in favor of reopening this PR.

**Pull request closed without merging**